### DISCUSSION TITLE
클로저에 대해 설명해 주세요

### DISCUSSION THREAD

**Discussion by JUDONGHYEOK** [woowacourse-study/2022-woowahan-bansanghwe#92] [JavaScript]
## 클로저란?
> 클로저는 넓은 의미로 함수와 함수가 선언된 Lexical Environment의 조합이다. 
> \- MDN [클로저](https://developer.mozilla.org/ko/docs/Web/JavaScript/Closures) - 

따라서 모든 함수가 클로저라고 말할 수 있다. 보통 함수는 렉시컬 스코프 즉 상위 스코프에 대한 참조는 함수가 정의가 평가되는 시점에 함수가 정의된 환경(위치)에 의해 결정되기 때문에 함수는 자신의 내부 슬롯 [[Environment]]에 자신이 정의된 환경 즉 상위 스코프의 참조를 저장한다. 클로저란 이러한 [[Environment]]와 이 함수 자체의 조합이다.

하지만 보통 우리가 클로저를 사용할 때는 위와 같이 말하진 않는다. 보통 클로저라고 하면 아래 함수와 같은 상황을 말한다.

![image](https://user-images.githubusercontent.com/65863017/190846448-a15b73c9-727a-49c1-bdb7-a3f4b95da464.png)

```javascript
const car = function(){
   let position = 0;
    return {
        move: function(){
            position++;
        },
        print: function(){
            console.log("현재 위치는 ",position," 입니다.")
        }
    }
}()

car.move();
car.print(); //  현재 위치는  1  입니다.

car.move();
car.print(); // 현재 위치는  2  입니다.
```

 위의 상황과 같이 보통 클로저는 내부 함수에서 외부 함수에 대한 변수를 참조하여 외부 함수의 실행 컨텍스트가 콜스택에서 제거된 후에도 내부 함수에서 외부 함수에 대한 변수를 즉 외부 함수의 Lexical Record를 참조하고 있기 때문에 해당 변수는 자바스크립트의 가비지컬렉터의 대상이 되질 않고 계속 남아있는 상황이다. 클로저는 보통 이러한 함수를 의미한다.

 위의 상황을 자세히 살펴보자. car 함수는 move와 print를 리턴하는 즉시 실행 함수이다. 즉 car에는 position을 참조하고 있는 함수를 갖는 move와 print 프로퍼티를 갖는 객체이다.  js에서 렉시컬 스코프를 따른다. 때문에 move프로퍼티 value와 print프로퍼티 value는 선언되는 시점에 스코프가 생성이 되고 이 스코프내에는 position변수가 참조되고 있기 때문에 position은 사라지지 않는다. 따라서 전역 실행 컨텍스트에서 move, print에서 positon이라는 자유 변수를 참조하고 있으므로 해당 익명함수의 실행 컨텍스트가 종료된 이후에서 해당 자유변수는 각각 move와 print의 함수에 scope내의 closure에서 참조하고 있기 때문에 가비지 컬렉터의 대상이 되지 않는다. 따라서 move와 print가 호출될 때도 해당 position 변수는 계속해서 메모리에 존재하는 모습을 확인할 수 있다.

## 그렇다면 클로저는 어떠한 상황에서 활용될 수 있을까? 
보통 클로저는 이러한 자유변수를 활용하기 위해 많이 사용된다. 해당 변수는 내부 함수의 스코프내에서만 존재하기 때문에 다른 컨텍스트의 스코프 체인에서는 접근이 불가능하다. 따라서 이러한 변수들은 자연스레 private화 되는 것이고 은닉화 할 수 있게 된다.

### 1. 디바운싱과 쓰로틀링
디바운싱과 쓰로틀링은 모두 API와 같이 무거운 태스크들을 연속해서 호출할 때의 과부하를 막기 위해서 사용된다. 디바운싱은 마지막 함수가 호출된 후 일정 시간이 지나기 전에 다시 호출되지 않도록 하는 것이고 쓰로틀링은 연이어 호출되는 함수들 중 마지막 함수(또는 제일 처음)만 호출하도록 하는 것이다. 
<img width="1121" alt="image" src="https://user-images.githubusercontent.com/65863017/190889038-f26848bc-c3e2-43e5-a157-14e290928090.png">
<img width="1121" alt="스크린샷 2022-09-18 오후 3 34 24" src="https://user-images.githubusercontent.com/65863017/190889086-164c0ddf-1bca-4105-a939-90062183e69e.png">
 이러한 디바운싱과 쓰로틀링을 활용할 때 내부 timerID를 클로저를 통해 자유변수화 시키면서 외부에서는 접근 불가능하게 막을 수 있다.

```javascript
function debounce(fn, delay) {
  let timer = null; // timer 은닉을 위해 클로저를 사용
  return function () {
    // 클로저 함수 안에서 this 와 arguments 변수로 디바운싱 함수의 스코프와 변수에 접근
    const context = this;
    const args = arguments;
    clearTimeout(timer); // delay 함수가 다시 호출되면 타이머를 리셋한다.
    timer = setTimeout(function () {
      fn.apply(context, args);
    }, delay); // delay 동안 호출이 없으면 fn을 실행한다.
  };
}
```

```javascript
function throttle(fn, threshhold = 250) {
  let lastCallTime, deferTimer;
  return function () {
    const context = this;
    const args = arguments;
    const now = Date.now();

    if (lastCallTime && now < lastCallTime + threshhold) {
      // 호출된 적이 있다면, 마지막 호출 시점부터 threshhold 만큼 시간이 지났을 때만 재호출
      clearTimeout(deferTimer);
      deferTimer = setTimeout(function () {
        lastCallTime = now;
        fn.apply(context, args);
      }, threshhold);
    } else {
      // 한번도 호출된 적 없으면 즉시 호출
      lastCallTime = now;
      fn.apply(context, args);
    }
  };
}
```

참고자료
https://developer.mozilla.org/ko/docs/Web/JavaScript/Closures
https://yoonho-devlog.tistory.com/148

**Comment by al-bur** [comment]
클로저에 대해 설명해주세요

**Comment by al-bur** [comment]
자유변수란?

**Comment by al-bur** [comment]
클로저 자체를 사용하신 적이 있나요?

**Comment by al-bur** [comment]
렉시컬 스코프란?

**Comment by al-bur** [comment]
스코프와 클로저의 차이는 뭔가요?

**Comment by al-bur** [comment]
스코프 체인이란 뭘까요?

**Comment by al-bur** [comment]
왜 스코프 체인이 일어나도록 언어가 설계 되었을까?

**Comment by al-bur** [comment]
스코프 체인의 장점?

**Comment by al-bur** [comment]
클로저라는 기법이 왜 생겨났다고 생각하나요?

**Comment by al-bur** [comment]
함수형 프로그램은 정확히 어떤 것을 의미하는 건가요?

**Comment by al-bur** [comment]
즉시 실행 함수와 클로저의 차이

**Comment by al-bur** [comment]
모듈과 클로저가 비슷하게 동작을 하는데, ES6 모듈이 나왔는데, 클로저를 중요하게 학습해야할 이유가 있을까요?

**Comment by al-bur** [comment]
은닉화가 뭘까요?

**Comment by al-bur** [comment]
`관심사의 분리`말고도 은닉화를 사용하는 이유?

**Comment by al-bur** [comment]
디바운싱 vs 쓰로틀링

**Comment by al-bur** [comment]
디바운싱과 쓰로틀링을 클로저를 이용하지 않고 구현할 수 있는 방법은?

**Comment by al-bur** [comment]
객체와 클래스의 인스턴스의 차이는?

**Comment by al-bur** [comment]
객체 지향이 뭘까요?

**Comment by al-bur** [comment]
프로젝트를 하시면서 객체 지향을 적용해본 적이 있나요?

**Comment by al-bur** [comment]
재사용성이 높다는 기준은 어떤 걸까요?

**Comment by al-bur** [comment]
관심사의 분리는 어떤식으로 했죠?

**Comment by al-bur** [comment]
관심사 분리와 선언형 프로그래밍은 어떤 지점에서 닿아있을까요?
